src/controllers/auth.js

```diff
@@ -0,0 +1,397 @@
+'use strict';
+
+const crypto = require('crypto');
+const { v4: uuidv4 } = require('uuid');
+const passport = require('passport');
+
+const { ApiError } = require('../exceptions/ApiError');
+const usersService = require('../services/users');
+const tokenService = require('../services/token');
+const jwtService = require('../services/jwt');
+const emailService = require('../services/email');
+const {
+  validateUsername,
+  validateEmail,
+  validatePassword,
+  validateEmailChange,
+  validateUserPassword,
+} = require('../utils/authValidators');
+
+async function register(req, res, next) {
+  const salt = crypto.randomBytes(16).toString('hex');
+
+  const {
+    username,
+    email,
+    password,
+  } = req.body;
+
+  const errors = {
+    username: validateUsername(username),
+    email: validateEmail(email),
+    password: validatePassword(password),
+  };
+
+  if (errors.username || errors.email || errors.password) {
+    throw ApiError.BadRequest('Validation error', errors);
+  }
+
+  const existedUser = await usersService.getOneByField({ email });
+
+  if (existedUser) {
+    throw ApiError.BadRequest('User with such email is already exists', {
+      email: 'User with such email is already exists',
+    });
+  }
+
+  crypto.pbkdf2(
+    password,
+    salt,
+    310000,
+    32,
+    'sha256',
+    (err, hashedPassword) => {
+      if (err) {
+        throw ApiError.UnexpectedError();
+      }
+
+      const activationToken = uuidv4();
+
+      usersService
+        .create({
+          username,
+          hashed_password: hashedPassword.toString('hex'),
+          salt,
+          email,
+          activationToken,
+        })
+        .then((user) => {
+          emailService.sendActivationLink(email, activationToken);
+          res.send({ user: usersService.normalize(user) });
+        })
+        .catch(error => {
+          throw ApiError.UnexpectedError(error);
+        });
+    },
+  );
+}
+
+async function patch(req, res, next) {
+  const userId = req.params.userId;
```

Comment:
```suggestion
  const { userId } = req.params;
```
check all other similar places